src/topology/algebra/module/character_space.lean

```diff
@@ -0,0 +1,98 @@
+/-
+Copyright (c) 2022 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import topology.algebra.module.weak_dual
+import algebra.algebra.spectrum
+
+/-!
+# Character space of a topological algebra
+
+The character space of a topological algebra is the subset of elements of the weak dual that
+are also algebra homomorphisms. This space is used in the Gelfand transform, which gives an
+isomorphism between a commutative C⋆-algebra and continuous functions on the character space
+of the algebra. This, in turn, is used to construct the continuous functional calculus on
+C⋆-algebras.
+
+
+## Implementation notes
+
+We define `character_space 𝕜 A` as a subset of the weak dual, which automatically puts the
+correct topology on the space. We then define `to_alg_hom` which provides the algebra homomorphism
+corresponding to any element. We also provide `to_clm` which provides the element as a
+continuous linear map. (Even though `weak_dual 𝕜 A` is a type copy of `A →L[𝕜] 𝕜`, this is
+often more convenient.)
+
+## TODO
+
+* Prove that the character space is a compact subset of the weak dual. This requires the
+  Banach-Alaoglu theorem.
+
+## Tags
+
+character space, Gelfand transform, functional calculus
+
+-/
+
+namespace weak_dual
+
+/-- The character space of a topological algebra is the subset of elements of the weak dual that
+are also algebra homomorphisms. -/
+def character_space (𝕜 : Type*) (A : Type*) [comm_semiring 𝕜] [topological_space 𝕜]
+  [has_continuous_add 𝕜] [has_continuous_const_smul 𝕜 𝕜]
+  [semiring A] [topological_space A] [module 𝕜 A] :=
+  {φ : weak_dual 𝕜 A | (φ 1 = 1) ∧ (∀ (x y : A), φ (x * y) = (φ x) * (φ y))}
```

Comment:
Alright, several comments about this:

1. Let's also handle the non-unital (and non-associative?) case simultaneously by replacing the condition that `φ 1 = 1` with `φ ≠ 0`. When `A` is unital and `𝕜` has no zero divisors, these are equivalent because the multiplicative condition guarantees `φ 1 = (φ 1) ^ 2`, so `(φ 1)(1 - φ 1) = 0`, so either`φ 1 = 0` (which by multiplicativity guarantees `φ = 0`), or else `φ 1 = 1`. So, it would be nice to have `φ ≠ 0` in the definition, and then a proof that `φ` is unital under the conditions described above.
2. I could potentially be convinced otherwise, but I think we actually want `character_space 𝕜 A` to be a subtype of `weak_dual 𝕜 A`. That way, we can actually have the type class `compact_space (character_space 𝕜 A)` (under appropriate hypotheses) instead of just `set.is_compact (character_space 𝕜 A)`. This will be nice when we consider `C(character_space 𝕜 A, 𝕜)` or related types. If you really want just the set separately, maybe we can call it `nonzero_characters` or something, and reserve `character_space` for the subtype?
3. I'm not sure how the algebraists will feel about us calling this `character_space` anyway, because multiplicative characters have a slightly different meaning in the theory of group representations. The `character_space` is sometimes also referred to in the literature by "maximal ideal space", but I really don't like that for us because we are not giving a formulation explicitly in terms of maximal ideals (even though under some assumptions they are equivalent).

Reply:
I forgot to mention: we just get the topology as the induced topology from the coercion. This isn't so unpleasant as the the image of the subtype is weak star closed, thus a set is closed in the subtype if and only if its coercion is closed in the weak dual.

Reply:
>     1. Let's also handle the non-unital (and non-associative?) case simultaneously by replacing the condition that `φ 1 = 1` with `φ ≠ 0`. When `A` is unital and `𝕜` has no zero divisors, these are equivalent because the multiplicative condition guarantees `φ 1 = (φ 1) ^ 2`, so `(φ 1)(1 - φ 1) = 0`, so either`φ 1 = 0` (which by multiplicativity guarantees `φ = 0`), or else `φ 1 = 1`. So, it would be nice to have `φ ≠ 0` in the definition, and then a proof that `φ` is unital under the conditions described above.

Ah, yes that makes sense, I'll do that.

>     2. I could potentially be convinced otherwise, but I think we actually want `character_space 𝕜 A` to be a subtype of `weak_dual 𝕜 A`. That way, we can actually have the type class `compact_space (character_space 𝕜 A)` (under appropriate hypotheses) instead of just `set.is_compact (character_space 𝕜 A)`. This will be nice when we consider `C(character_space 𝕜 A, 𝕜)` or related types. If you really want just the set separately, maybe we can call it `nonzero_characters` or something, and reserve `character_space` for the subtype?

My first instinct was also to go for a subtype, but it actually turned out to be less convenient in many respects, because you actually end up with a type copy of the subtype and not the subtype itself. So you have to put in the coercion to `weak_dual 𝕜 A` manually, the topological space instance also, etc. Whereas when the set gets lifted to a type these problems don't arise. Also I don't see what would be the problem with putting that compact space instance on it, I think that should just work.

>     3. I'm not sure how the algebraists will feel about us calling this `character_space` anyway, because multiplicative characters have a slightly different meaning in the theory of group representations. The `character_space` is sometimes also referred to in the literature by "maximal ideal space", but I really don't like that for us because we are not giving a formulation explicitly in terms of maximal ideals (even though under some assumptions they are equivalent).

I agree it sounds quite generic. Right now it's actually `weak_dual.character_space` which should be enough to avoid interference with other areas, but I'm open to better ideas.

Reply:
I don't understand the response to #2. Why do you get a type copy of the subtype and not the subtype itself? I mean, I was just thinking:
```lean
{φ : weak_dual 𝕜 A // (φ ≠ 0) ∧ (∀ (x y : A), φ (x * y) = (φ x) * (φ y))}
```
Yes, you have to create the topology from the induced topology of the coercion, but I don't understand all the other problems. 

I also don't understand how the `compact_space` instance would work. Right now `character_space 𝕜 A` has type `set (weak_dual 𝕜 A)`. So wouldn't `compact_space (character_space 𝕜 A)` say `is_compact (set.univ : set (character_space 𝕜 A))`? And then `set (character_space 𝕜 A) : set (set (weak_dual 𝕜 A))`? And this is not what you want, right?

Reply:
As for 3, I think the `weak_dual` namespace is sufficient disambiguation that no one should complain. I hadn't noticed that before.

Reply:
> I don't understand the response to 2. Why do you get a type copy of the subtype and not the subtype itself? I mean, I was just thinking:
> 
> ```lean
> {φ : weak_dual 𝕜 A // (φ ≠ 0) ∧ (∀ (x y : A), φ (x * y) = (φ x) * (φ y))}
> ```

If you try changing the current definition of `character_space` to a subtype, and then try the following:
```lean
example : has_coe (character_space 𝕜 A) (weak_dual 𝕜 A) := by apply_instance
example : has_coe {φ : weak_dual 𝕜 A // (φ ≠ 0) ∧ (∀ (x y : A), φ (x * y) = (φ x) * (φ y))} 
  (weak_dual 𝕜 A) := by apply_instance
```
you'll see that only the second one works, because `character_space` is not literally the subtype, but only a type copy. I think we would have to use notation to make it work properly.
 
> Yes, you have to create the topology from the induced topology of the coercion, but I don't understand all the other problems.
> 
> I also don't understand how the `compact_space` instance would work. Right now `character_space 𝕜 A` has type `set (weak_dual 𝕜 A)`. So wouldn't `compact_space (character_space 𝕜 A)` say `is_compact (set.univ : set (character_space 𝕜 A))`? And then `set (character_space 𝕜 A) : set (set (weak_dual 𝕜 A))`? And this is not what you want, right?

The compact space instance would be on `↥(character_space 𝕜 A)`, which is `character_space 𝕜 A` lifted to a type. If you unfold the definition, it's basically what you wrote, but for a subtype it would also be something similar.